### PR TITLE
GREAT INTEGRATION

### PR DESCRIPTION
**Pull request by DSsoto**
Lets make sure that everything is clean here before we merge back to master.

It would be helpful to have as many people as possible review this.

### CODE REVIEW

File: perception/mil_vision/great_merge_purgatory/README.txt [hunk 1]
```diff
@@ -0,0 +1,5 @@
+Some of these files can be made generic and therefore be able to stay as a part of mil_common. Until that is done, they will remain in this folder.
```
Comment (RustyBamboo) [inline perception/mil_vision/great_merge_purgatory/README.txt]: great_merge_purgatory

File: perception/mil_vision/include/mil_vision_lib/colorizer/camera_observer.hpp [hunk 2]
```diff
@@ -0,0 +1,45 @@
+#pragma once
+
+#include <mil_vision_lib/colorizer/common.hpp>  // Common includes are here
+#include <mil_vision_lib/colorizer/color_observation.hpp>
+#include <tf/transform_listener.h>
+#include <image_geometry/pinhole_camera_model.h>
+#include <opencv2/core/eigen.hpp>
+
+namespace nav{
```
Comment (kev-the-dev) [inline perception/mil_vision/include/mil_vision_lib/colorizer/camera_observer.hpp]: All namespaces should be ```mil_vision``` for resources in this library.

File: perception/mil_vision/include/mil_vision_lib/colorizer/camera_observer.hpp [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  
+class CameraObserver
```
Comment (kev-the-dev) [inline perception/mil_vision/include/mil_vision_lib/colorizer/camera_observer.hpp]: Let's have just one (David's) wrapper for camera subscribers/publishers. @DSsoto can you remove the other ones that do the same thing?

File: perception/mil_vision/include/mil_vision_lib/cv_tools.hpp [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+void statistical_image_segmentation(const cv::Mat &src, cv::Mat &dest,
+                                    cv::Mat &debug_img, const int hist_size,
+                                    const float **ranges, const int target,
+                                    std::string image_name = "Unnamed Image",
+                                    bool ret_dbg_img = false,
+                                    const float sigma = 1.5,
+                                    const float low_thresh_gain = 0.5,
+                                    const float high_thresh_gain = 0.5);
+
+cv::Mat triangulate_Linear_LS(cv::Mat mat_P_l, cv::Mat mat_P_r,
```
Comment (kev-the-dev) [inline perception/mil_vision/include/mil_vision_lib/cv_tools.hpp:95]: No documentation for this function. I do I use it/ what does it do? Same with others

File: perception/mil_vision/include/mil_vision_lib/point_cloud_algorithms.hpp [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+  PCD output_pcd;
+  ros::NodeHandle nh;
+  ros::Subscriber cloud_sub;
+  ros::Publisher cloud_pub;
+  std::string input_pcd_topic;
+  std::string output_pcd_topic;
+};
+
+
+class PcdColorizer : public PcdSubPubAlgorithm{
```
Comment (kev-the-dev) [inline perception/mil_vision/include/mil_vision_lib/point_cloud_algorithms.hpp:36]: Add TODO comment, mention unfinished

File: perception/mil_vision/include/mil_vision_lib/visualization.hpp [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+namespace nav {
+
+// ******* 3D Visualization *******
+class RvizVisualizer {
+ public:
+  ros::Publisher rviz_pub;
+  ros::NodeHandle nh;
+  RvizVisualizer(std::string rviz_topic);
+  void visualize_buoy(geometry_msgs::Pose &pose, std::string &frame_id);
```
Comment (kev-the-dev) [inline perception/mil_vision/include/mil_vision_lib/visualization.hpp]: sub specific, remove

File: perception/mil_vision/object_classification/lidar_to_image.py [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+import genpy
+import cv2
+import navigator_tools as nt
+from navigator_msgs.srv import ObjectDBQuery, ObjectDBQueryRequest
+from image_geometry import PinholeCameraModel
+import numpy as np
+___author___ = "Tess Bianchi"
+
+
+class LidarToImage(object):
```
Comment (kev-the-dev) [inline perception/mil_vision/object_classification/lidar_to_image.py]: Specific to Navigator, refactor or remove

File: perception/mil_vision/object_classification/object_classification.py [hunk 8]
```diff
@@ -0,0 +1,150 @@
+
+from roi_generator import ROI_Collection
+from navigator_tools import CvDebug, BagCrawler
```
Comment (kev-the-dev) [inline perception/mil_vision/object_classification/object_classification.py]: Navigator dependency, update to mil_common 

Specific to Navigator, refactor or remove

File: perception/mil_vision/object_classification/roi_generator.py [hunk 9]
```diff
@@ -0,0 +1,193 @@
+#!/usr/bin/python
+from navigator_tools import BagCrawler
```
Comment (kev-the-dev) [inline perception/mil_vision/object_classification/roi_generator.py]: object_classification folder should be removed or refactored to be generic

File: perception/mil_vision/package.xml [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+  <run_depend>std_msgs</run_depend>
+  <build_depend>geometry_msgs</build_depend>
+  <run_depend>geometry_msgs</run_depend>
+  <build_depend>sensor_msgs</build_depend>
+  <run_depend>sensor_msgs</run_depend>
+  <run_depend>libpcl-all</run_depend>
+  <build_depend>tf2_sensor_msgs</build_depend>
+  <run_depend>tf2_sensor_msgs</run_depend>
+  <build_depend>tf2_geometry_msgs</build_depend>
+  <build_depend>navigator_tools</build_depend>
```
Comment (kev-the-dev) [inline perception/mil_vision/package.xml]: mil_common should not depened on navgiator

File: perception/mil_vision/ros_tools/easy_thresh.py [hunk 11]
```diff
@@ -0,0 +1,162 @@
+#!/usr/bin/python
```
Comment (kev-the-dev) [inline perception/mil_vision/ros_tools/easy_thresh.py:1]: Broken for me on kinetic, someone else try to run it

File: perception/mil_vision/ros_tools/on_the_fly_thresholder.py [hunk 12]
```diff
@@ -0,0 +1,117 @@
+#!/usr/bin/python
+from txros import util, NodeHandle
+from twisted.internet import defer, reactor
+from sensor_msgs.msg import Image, CompressedImage
+from cv_bridge import CvBridge, CvBridgeError
+
+from navigator_tools import image_helpers
```
Comment (kev-the-dev) [inline perception/mil_vision/ros_tools/on_the_fly_thresholder.py]: use mil_common instead

File: perception/mil_vision/setup.py [hunk 13]
```diff
@@ -0,0 +1,11 @@
+# ! DO NOT MANUALLY INVOKE THIS setup.py, USE CATKIN INSTEAD
+
+from distutils.core import setup
+from catkin_pkg.python_setup import generate_distutils_setup
+
+# fetch values from package.xml
+setup_args = generate_distutils_setup(
+    packages=[''],
```
Comment (kev-the-dev) [inline perception/mil_vision/setup.py:8]: Should this be empty? Can you import the python stuff from this package?

File: perception/mil_vision/src/mil_vision_lib/active_contours.cpp [hunk 14]
```diff
@@ -0,0 +1,253 @@
+#include <mil_vision_lib/active_contours.hpp>
+
+using namespace std;
+using namespace cv;
+
+namespace nav
+{
+
+namespace Perturbations
```
Comment (kev-the-dev) [inline perception/mil_vision/src/mil_vision_lib/active_contours.cpp]: Documentation

File: perception/mil_vision/src/mil_vision_lib/colorizer/camera_observer.cpp [hunk 15]
```diff
@@ -0,0 +1,73 @@
+#include <colorizer/camera_observer.hpp>
+
+namespace nav{
+
+using mil::tools::operator "" _s; // converts to std::string
+  
+CameraObserver::CameraObserver(ros::NodeHandle &nh, std::string &pcd_in_topic, std::string &cam_topic, size_t hist_size)
```
Comment (kev-the-dev) [inline perception/mil_vision/src/mil_vision_lib/colorizer/camera_observer.cpp]: Document. What does this do?

File: perception/mil_vision/src/mil_vision_lib/colorizer/color_observation.cpp [hunk 16]
```diff
@@ -0,0 +1,15 @@
+#include <colorizer/color_observation.hpp>
```
Comment (kev-the-dev) [inline perception/mil_vision/src/mil_vision_lib/colorizer/color_observation.cpp:1]: Document the colorizer package. Add a readme

File: perception/mil_vision/src/mil_vision_lib/colorizer/pcd_colorizer.cpp [hunk 17]
```diff
@@ -0,0 +1,173 @@
+#include <mil_vision_lib/colorizer/pcd_colorizer.hpp>
+
+using namespace std;
+using namespace cv;
+
+namespace nav{
```
Comment (kev-the-dev) [inline perception/mil_vision/src/mil_vision_lib/colorizer/pcd_colorizer.cpp]: namespace

File: perception/mil_vision/src/mil_vision_lib/cv_utils.cc [hunk 18]
```diff
@@ -0,0 +1,656 @@
+#include <mil_vision_lib/cv_tools.hpp>
```
Comment (kev-the-dev) [inline perception/mil_vision/src/mil_vision_lib/cv_utils.cc:1]: Rename to cpp

Document these functions.

File: utils/mil_msgs/CMakeLists.txt [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+    actionlib_msgs
+    cmake_modules
+)
+catkin_python_setup()
+
+add_action_files(FILES
+  MoveTo.action
+)
+
+add_message_files(FILES
```
Comment (kev-the-dev) [inline utils/mil_msgs/CMakeLists.txt:18]: Some of these exist in geometry_msgs/other core ros packages.

File: utils/mil_msgs/CMakeLists.txt [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+    actionlib
+    interactive_markers
+    std_msgs
+    actionlib_msgs
+    cmake_modules
+)
+catkin_python_setup()
+
+add_action_files(FILES
+  MoveTo.action
```
Comment (kev-the-dev) [inline utils/mil_msgs/CMakeLists.txt]: Is this vehicle dependent?

File: utils/mil_msgs/include/mil_msgs/msg_helpers.h [hunk 21]
```diff
@@ -0,0 +1,77 @@
+#pragma once
+
+#include <Eigen/Dense>
+#include <tf/tf.h>
+
+namespace uf_common {
```
Comment (kev-the-dev) [inline utils/mil_msgs/include/mil_msgs/msg_helpers.h]: namespace

File: utils/mil_msgs/include/mil_msgs/msg_helpers.h [hunk 22]
```diff
@@ -0,0 +1,77 @@
+#pragma once
```
Comment (kev-the-dev) [inline utils/mil_msgs/include/mil_msgs/msg_helpers.h]: msg_helpers should be moved to dev tools package. This package should be just message/service/action definitions common to mil projects

File: utils/mil_msgs/scripts/param_saver [hunk 23]
```diff
@@ -0,0 +1,40 @@
+#!/usr/bin/env python
+
+import os
+import random
+import yaml
+
+import roslib
+roslib.load_manifest('uf_common')
```
Comment (kev-the-dev) [inline utils/mil_msgs/scripts/param_saver]: don't need with catkin. Also these scripts should go to tools package

File: utils/mil_tools/include/mil_tools/mil_tools.hpp [hunk 24]
```diff
@@ -0,0 +1,33 @@
+#pragma once
+
+#include <vector>
+#include <string>
+#include <algorithm>
+#include <ros/ros.h>
+#include <opencv2/core/core.hpp>
+#include <opencv2/imgproc/imgproc.hpp>
+
+namespace mil{
```
Comment (kev-the-dev) [inline utils/mil_tools/include/mil_tools/mil_tools.hpp]: rename to mil_tools

File: utils/mil_tools/mil_misc_tools/download.py [hunk 25]
```diff
@@ -0,0 +1,57 @@
...
+    html = response.read()
+    if output_filename is not None:
+        f = open(output_filename, 'w')
+        f.write(html)
+        f.close()
+    return html
+
+
+if __name__ == '__main__':
+    sub_model_url = "http://goo.gl/f0ennf?gdriveurl"
```
Comment (kev-the-dev) [inline utils/mil_tools/mil_misc_tools/download.py:56]: bad

File: utils/mil_tools/mil_ros_tools/bag_crawler.py [hunk 26]
```diff
@@ -0,0 +1,71 @@
...
+This file wis written by the team at UF MIL for the 2016 robosub competition.
+
+github.com/uf-mil
+"""
+import rosbag
+from cv_bridge import CvBridge
+import tqdm
+
+
+class BagCrawler(object):
```
Comment (kev-the-dev) [inline utils/mil_tools/mil_ros_tools/bag_crawler.py:13]: Document

File: utils/mil_tools/mil_ros_tools/func_helpers.py [hunk 27]
```diff
@@ -0,0 +1,38 @@
...
+
+if __name__ == '__main__':
+
+    print add(1, 2)
+    print add(1, 2)
+
+    print add(2, 3)
+    print add(2, 3)
+
+    print gooberstein("hello")
```
Comment (kev-the-dev) [inline utils/mil_tools/mil_ros_tools/func_helpers.py:36]: wat

File: utils/mil_tools/mil_ros_tools/func_helpers.py [hunk 28]
```diff
@@ -0,0 +1,38 @@
...
+        if args not in self.call_dict.keys():
+            result = self.func(*args)
+            self.call_dict[args] = result
+        else:
+            result = self.call_dict[args]
+        return result
+
+
+@Cache
+def add(a, b):
```
Comment (kev-the-dev) [inline utils/mil_tools/mil_ros_tools/func_helpers.py:17]: useful

File: utils/mil_tools/mil_ros_tools/image_helpers.py [hunk 29]
```diff
@@ -0,0 +1,116 @@
...
+'''
+import rospy
+import numpy as np
+from os import path
+from cv_bridge import CvBridge, CvBridgeError
+from sensor_msgs.msg import Image, CameraInfo
+from mil_ros_tools import wait_for_param
+
+
+def get_parameter_range(parameter_root):
```
Comment (kev-the-dev) [inline utils/mil_tools/mil_ros_tools/image_helpers.py:15]: sub specific, remove

File: utils/mil_tools/mil_ros_tools/image_helpers.py [hunk 30]
```diff
@@ -0,0 +1,116 @@
...
+        self.last_image = None
+        self.im_sub = rospy.Subscriber(topic, Image, self.convert, queue_size=queue_size)
+
+        root_topic, image_subtopic = path.split(topic)
+        self.info_sub = rospy.Subscriber(root_topic + '/camera_info', CameraInfo, self.info_cb, queue_size=queue_size)
+
+        self.bridge = CvBridge()
+        self.callback = callback
+
+    def wait_for_camera_info(self, timeout=10):
```
Comment (kev-the-dev) [inline utils/mil_tools/mil_ros_tools/image_helpers.py:84]: known bug when time goes backwards, fixed on sub8 version

File: utils/mil_tools/mil_ros_tools/msg_helpers.py [hunk 31]
```diff
@@ -0,0 +1,194 @@
+import numpy as np
+from tf import transformations
+import geometry_msgs.msg as geometry_msgs
+import std_msgs.msg as std_msgs
+import nav_msgs.msg as nav_msgs
+import rospy
+
+
+def rosmsg_to_numpy(rosmsg, keys=None):
```
Comment (kev-the-dev) [inline utils/mil_tools/mil_ros_tools/msg_helpers.py:9]: @RustyBamboo has some fix for this. Please PR it

File: utils/mil_tools/nodes/tf_fudger.py [hunk 32]
```diff
@@ -0,0 +1,172 @@
...
+    if k == ord('q'):
+        q_mode = not q_mode
+
+    if k == ord('r'):
+        reset()
+
+    if k == ord('z'):
+        set_bars((0, 0, 0), (0, 0, 0))
+
+    # This functionality was configured to replace lines in navigator's tf launch file, should refactor to be general later
```
Comment (kev-the-dev) [inline utils/mil_tools/nodes/tf_fudger.py:125]: IF someone wants to refactor to be generic, that'd be cool

File: utils/mil_tools/nodes/video_player [hunk 33]
```diff
@@ -0,0 +1,134 @@
...
+      self.roi_height = d if d > 0 else 1
+
+  def roi_width_cb(self, d):
+      self.roi_width = d if d > 0 else 1
+
+  def pause(self):
+      self.paused = not self.paused
+
+def main():
+    rospy.init_node('video_player', anonymous=True)
```
Comment (kev-the-dev) [inline utils/mil_tools/nodes/video_player:129]: use argparse instead of params

**Review by kev-the-dev — CHANGES_REQUESTED** [review]
Could use some work. 6/10

Some of these changes need to be done before merging:

* namespace changes
* anything specific to sub8/navigator

Later
* document

**Comment by kev-the-dev** [comment]
Comments addressed and tested (at least for building) with SubjuGator, except for changes mentioned in #30, which can be done in a post-integration repo.

**Review by kev-the-dev — APPROVED** [review]

**Comment by sentree** [comment]
This built on the CI and passed all unit tests except the ros_alarms test that always fails.
-- Your friendly temporary CI guy

**Pull request merged**